taichi/program/state_flow_graph.cpp

```diff
@@ -101,77 +95,96 @@ void StateFlowGraph::insert_state_flow(Node *from, Node *to, AsyncState state) {
-      // and node_b that writes the list (the following might be too
-      // conservative)
-      auto a_has_path_to_b = has_path[i] & has_path_reverse[j];
-      a_has_path_to_b[i] = a_has_path_to_b[j] = false;
-      // Test if there is a path node_a->node_c->node_b here
-      if (a_has_path_to_b.any())
-        continue;
-      */
+      bool erasing = false;
 
-      TI_INFO("Common list generation {} and {}", node_a->string(),
-              node_b->string());
-      common_pairs.emplace_back(std::make_pair(i, j));
+      for (int j = i + 1; j < listgens.size(); j++) {
+        auto node_b = listgens[j];
+
+        // Test if two list generations share the same mask and parent list
+        auto snode = node_a->meta->snode;
+
+        auto mask_state = AsyncState{snode, AsyncState::Type::mask};
+        auto list_state = AsyncState{snode, AsyncState::Type::list};
+        auto parent_list_state =
+            AsyncState{snode->parent, AsyncState::Type::list};
+
+        TI_ASSERT(node_a->input_edges[mask_state].size() == 1);
+        TI_ASSERT(node_b->input_edges[mask_state].size() == 1);
+
+        if (*node_a->input_edges[mask_state].begin() !=
+            *node_b->input_edges[mask_state].begin())
+          break;
+
+        TI_ASSERT(node_a->input_edges[parent_list_state].size() == 1);
+        TI_ASSERT(node_b->input_edges[parent_list_state].size() == 1);
+        if (*node_a->input_edges[parent_list_state].begin() !=
+            *node_b->input_edges[parent_list_state].begin())
+          break;
+
+        TI_ASSERT(node_b->input_edges[list_state].size() == 1);
+        Node *clear_node = *node_b->input_edges[list_state].begin();
+
+        const OffloadedStmt *clear_node_offload =
+            clear_node->rec.ir_handle.ir()->as<OffloadedStmt>();
+        TI_ASSERT(clear_node_offload->body->statements.size() == 1);
+        TI_ASSERT(clear_node_offload->body->statements[0]->is<ClearListStmt>());
+
+        // erase the serial task containing ClearListStmt
+        nodes_to_delete.insert(clear_node->node_id);
+
+        TI_DEBUG("Common list generation {} and (to erase) {}",
+                 node_a->string(), node_b->string());
+
+        nodes_to_delete.insert(node_b->node_id);
+        erasing = true;
+      }
+
+      if (erasing)
+        break;
     }
   }
 
-  std::unordered_set<int> nodes_to_delete;
-  // Erase node j
-  // Note: the corresponding ClearListStmt should be removed in later DSE passes
-  for (auto p : common_pairs) {
-    auto i = p.first;
-    auto j = p.second;
-    TI_INFO("Eliminating {}", nodes_[j]->string());
-    replace_reference(nodes_[j].get(), nodes_[i].get(),
-                      /*only_output_edges=*/true);
+  TI_ASSERT(nodes_to_delete.size() % 2 == 0);
+
+  if (!nodes_to_delete.empty()) {
     modified = true;
-    nodes_to_delete.insert(j);
+    delete_nodes(nodes_to_delete);
+    // Note: DO NOT topo sort the nodes here. Node deletion destroys order
+    // independency.
```

Comment:
@xumingkuan I have observed running `topo_sort` on an intermediate SFG (which doesn't have the properties we assumed) can lead to a wrong execution order of tasks. I'm not sure if this is also an issue in `fuse()`.

Reply:
In `fuse()`, when I merge node y into node x, I replace all edges from/to node y with edges from/to node x. I think in this way, we should have all state-flow chains staying correct?

Reply:
OK, that sounds reasonable - rebuilding the graph using the old topological order sounds more trivially correct though :-)

Reply:
The tricky point is that, the old topological order may not be correct -- consider the following case:
```
   A
  / \
 /   \
 V   V
 B   D
 |   |
 |   |
 V   V
 C   E
```
`ABCDE` is a correct topological order. However, if we fuse task E into B, `A(BE)CD` is not a correct topological order -- the only correct order is `AD(BE)C`.

Reply:
Interesting! It seems that merging nodes has a different correct correctness constraint compared to node deletion :-) Thanks for pointing that out.